### ISSUE TITLE
How do I get hdevtools to check files within stack projects?

### ISSUE THREAD

**Issue by flounders** [hdevtools/hdevtools#25]
I don't know when exactly this took place. Within the last couple weeks I have noticed this error message when Neomake invokes hdevtools: `Neomake: hdevtools completed with exit code 1`. I decided to do something about it today. I figured it might have something to do with GHC 8.0.1. So I've reinstalled hdevtools with stack and the resolver is nightly-2016-06-21. It works just fine on my XMonad config and even shows a warning about a shadow binding in my file. Anytime I try to check any of my files in a stack project it seems to have issues.

If I run hdevtools without --no-ghc-package-path it prints `Cabal error: ghc-pkg dump failed`. I killed the server and tried running check to which I am greeted with `hdevtools: <socket: 4>: hGetLine: end of file`. I've tried using absolute paths to check, src/Main.hs and -g -isrc Main.hs which results in the same thing.

Here is the direct output from my shell:

```
~/workspace/learning/battleship > stack exec -- hdevtools check src/Main.hs
Cabal error: ghc-pkg dump failed
~/workspace/learning/battleship > stack exec hdevtools -- --status
Server is running.
~/workspace/learning/battleship > stack exec hdevtools -- --stop-server
Shutting down server.
~/workspace/learning/battleship > stack exec --no-ghc-package-path hdevtools -- check src/Main.hs
hdevtools: <socket: 4>: hGetLine: end of file
~/workspace/learning/battleship > stack exec hdevtools -- --status
hdevtools: connect: does not exist (No such file or directory)
~/workspace/learning/battleship > stack exec --no-ghc-package-path hdevtools -- --start-server
~/workspace/learning/battleship > stack exec --no-ghc-package-path hdevtools -- --status                                                                                                                                          
Server is running.
~/workspace/learning/battleship > stack exec --no-ghc-package-path hdevtools -- check src/Main.hs                                                                                                                                 
hdevtools: <socket: 3>: hGetLine: end of file
~/workspace/learning/battleship > stack exec --no-ghc-package-path hdevtools -- check -g -irsc src/Main.hs
hdevtools: <socket: 4>: hGetLine: end of file
~/workspace/learning/battleship > stack exec --no-ghc-package-path hdevtools -- check -g -irsc Main.hs    
hdevtools: <socket: 4>: hGetLine: end of file
~/workspace/learning/battleship > stack exec --no-ghc-package-path hdevtools -- check /home/swilliams/workspace/learning/battleship/src/Main.hs 
hdevtools: <socket: 4>: hGetLine: end of file
```

Here's what I'm using and versions:

1. Neovim 0.1.4 (and Neomake)
2. GHC 8.0.1 (System wide for XMonad)
3. stack 1.1.2
4. hdevtools 0.1.3.2 (ghc-8.0.1-x86-64-linux, cabal-1.24.0.0)
5. Arch Linux x86_64 (updated on June 23 2016 in the morning Eastern Daylight Time)

Also something else I'm wondering about is if there is a place on IRC, mailing-list, forum or any other appropriate means for addressing user configuration issues because I'm guessing it's my fault and not the program.

**Comment by ch1bo** [comment]
Hey,

currently Github issues is the way I try to support people on such topics.

First of all, hdevtools spawns a server if no one is running, and keeps it alive until the client is called with --stop-server or the unix domain socket (created in the working dir by default) is deleted. So if you try to run hdevtools in different configurations / versions, always make sure you start a fresh server.

Next, hdevtools has to be compiled with the same ghc version as the modules it loads / checks. The hGetLine error is in the client when the server is not responding correctly, you can probably get an error message by startng the server in non-daemon mode with `--start-server -n.

So, stop the server, maybe re-install hdevtools with the stack resolver of your project and have a look whether you get server errors.

Btw, today I fixed an issue in finding the stack.yaml, maybe you could also try the cloned master branch?

**Comment by flounders** [comment]
Outside of stack projects it's working fine with stackage build that has GHC 8.0.1. Otherwise I get this message when I run hdevtools in non-daemon mode: `hdevtools: /usr/lib/ghc-7.10.3/settings: openFile: does not exist (No such file or directory)` If I run the hdevtools that I have built with 8.0.1 inside a stack project I get the same thing. I have tried using the same resolver as I have on my project and that doesn't seem to make a difference.

Based on what you have said regarding hdevtools needing to be compiled with the same version of GHC as my projects does that mean I will need to have sandboxed versions of it in all of my projects if the resolvers differ?

I did try the master branch and it doesn't seem to have any effect.

**Comment by ch1bo** [comment]
Using the same resolvers (or installing hdevtools again from within your project) just ensures it is compiled against the same GHC as your project is using. 

So in order to use the same GHC it has to be invoked via stack as well (as stack knows where GHC versions installed by it reside - i.e. `~/.stack/programs`). Have a look at the alias I added to the readme. Does that help?

**Comment by flounders** [comment]
I've been running the command that is aliased which is `stack exec --no-ghc-package-path hdevtools --` and it seems to be running into `/usr/lib/ghc-{version}/settings: openFile: does not exist` with GHC versions prior to 8.0.1 if I start the server within the stack environment using no-daemon mode, or I just get the opaque end of line message otherwise. The issue seems to be in it finding that settings file for older versions. I just set it as an alias to be sure and that hasn't changed anything except when I run it without no-daemon mode it just hangs and never tells me why it isn't working. Attempting to start the server manually with the -n flag doesn't produce an error until you go to check status and it says `connect: does not exist`. How do I setup things up for it to see the settings file in ~/.stack/programs/?

**Comment by ch1bo** [comment]
Hm..never needed to point to settings files, probably some mismatch of ghc-pkg and package databases. Any concrete steps on how to reproduce your problem? Is the project open source? Which resolver? From where is hdevtools invoked?

**Comment by flounders** [comment]
I just checked with one of my other projects and it gets the same error messages plus this:

```
stack exec --no-ghc-package-path hdevtools -- --start-server -n
Run from outside a project, using implicit global project config
Compiler version mismatched, found ghc-7.10.3 (x86_64), but expected minor version match with ghc-8.0.1 (x86_64) (based on resolver setting in /home/swilliams/.stack/global-project/stack.yaml).
Try running "stack setup" to install the correct GHC into /home/swilliams/.stack/programs/x86_64-linux/
hdevtools: /usr/lib/ghc-7.10.3/settings: openFile: does not exist (No such file or directory)
```

This is invoked in my project directory for [LambdaPass](https://github.com/flounders/lambdaPass). The directory I've been trying this in is for another project I haven't put on GitHub yet.

Here are the steps:

1. Build and install git clone of hdevtools using lts-5.13 resolver and `stack install`.
2. Change into my LambdaPass project directory at ~/workspace/active_proects/LambdaPass
3. Run `stack exec --no-ghc-package-path hdevtools -- --start-server -n` in the LambdaPass base project directory.

These steps result in the input I pasted above.

**Comment by ch1bo** [comment]
So your global config (which was used in your first invocation) defaults to a ghc-8.0.1 while lts-5.13 (ghc-7.10.3) was used to build hdevtools? 

This could explain an error of the GHC library looking for 7.10.3/settings in a 8.0.1 environment.

But you describe steps in which hdevtools also not works from within LamdaPass, which is built with lts-5.13, am I right? I will try to reproduce this tomorrow.

**Comment by flounders** [comment]
I just changed the resolver to lts-5.13 for global and it didn't seem to have an effect. You are correct regarding LambdaPass.

**Comment by ch1bo** [comment]
Alright, could even reproduce the error with the `hdevtools` repo itself. Seems like the `ghc-paths` module provides the `/usr/lib/ghc-7.10.3` path (hardcoded into it) and when distros now move to system wide ghc-8.0.1 this path is obviously wrong. I pushed a fix to branch `stack-ghc-libdir` if you like to try it.

Thanks for reporting by the way!

**Comment by flounders** [comment]
You're welcome. It seems to be working fine on the command-line now. It looks like Neomake is still having issues with it, but I guess it isn't wrapping the command in stack exec, so I should be able to straighten that out on my own.

**Comment by ch1bo** [comment]
May I ask you to contribute a section about neomake in the README.md if you get it running? :)

**Comment by flounders** [comment]
Yeah I'd be happy to contribute. It may take me a bit to figure out how plugins and Vim script work first, though.

**Comment by bonds** [comment]
I'm working on neomake + stack + hdevtools support here: https://github.com/neomake/neomake/pull/522

**Comment by ch1bo** [comment]
Just recently an issue with neomake has  been resolved. I don't use it though. 

Anyway I think that the necessity of running hdevtoola via stack exec can be resolved easily, which should streamline use alot.

**Comment by ch1bo** [comment]
ping @flounders @bonds
I just fixed stack support such that it is NOT required to run hdevtools via `stack exec` anymore. If you fancy you can try that branch and provide some feedback. Thanks in advance!

Hopefully we can resolve this issue once that branch is merged.

**Comment by flounders** [comment]
I just built the latest commit from master, and tried hdevtools --start-server-n and it's giving the following: 

```
Run from outside a project, using implicit global project config
hdevtools: /usr/lib/ghc-7.10.3/settings: openFile: does not exist (No such file or directory)
```

Same result whether I run it in lambdaPass or the hdevtools repo itself.

Trying the same with `stack exec`  yields the same.

**Comment by ch1bo** [comment]
Sorry forgot to mention that I have not yet merged into master and the improvement should still be on `stack-ghc-libdir` branch.

**Comment by flounders** [comment]
Seems to be working perfectly. Nice work.

**Comment by ch1bo** [comment]
Cool, btw its released as version 0.1.4.0
